snow/engine/avalanche/bootstrap/bootstrapper.go

```diff
@@ -346,24 +358,54 @@ func (b *Bootstrapper) ForceAccepted(acceptedContainerIDs []ids.ID) error {
 	return b.process(toProcess...)
 }
 
-// Finish bootstrapping
-func (b *Bootstrapper) finish() error {
+// checkFinish repeatedly executes pending transactions and requests new frontier blocks until there aren't any new ones
+// after which it finishes the bootstrap process
+func (b *Bootstrapper) checkFinish() error {
 	// If there are outstanding requests for vertices or we still need to fetch vertices, we can't finish
 	if b.Ctx.IsBootstrapped() || b.OutstandingRequests.Len() > 0 || b.needToFetch.Len() > 0 {
 		return nil
 	}
 
 	b.Ctx.Log.Info("bootstrapping fetched %d vertices. executing transaction state transitions...",
 		b.NumFetched)
-	if err := b.executeAll(b.TxBlocked, b.Ctx.DecisionDispatcher); err != nil {
+
+	_, err := b.executeAll(b.TxBlocked, b.Ctx.DecisionDispatcher)
+	if err != nil {
 		return err
 	}
 
 	b.Ctx.Log.Info("executing vertex state transitions...")
-	if err := b.executeAll(b.VtxBlocked, b.Ctx.ConsensusDispatcher); err != nil {
+	executedTxs, err := b.executeAll(b.VtxBlocked, b.Ctx.ConsensusDispatcher)
+	if err != nil {
 		return err
 	}
 
+	// exec - 10k ; prev - 0     ; runs again because prev is math.MaxInt32
+	// exec - 20  ; prev - 10k/2 ; runs again
+	// exec - 2   ; prev - 20/2  ; runs again
+	// exec - 1   ; prev - 2/1   ; doesnt run again
+	//
+	// exec - 0   ; prev - 0     ; stops running as there's nothing fetched
+	previousStateTransitionsThreshold := b.executedStateTransitions
+	if previousStateTransitionsThreshold == 0 {
+		previousStateTransitionsThreshold = math.MaxInt32
+	} else {
+		previousStateTransitionsThreshold /= 2
+	}
```

Comment:
Why does this exist? We should be _initializing_ `b.executedStateTransitions` in `Initialize`.